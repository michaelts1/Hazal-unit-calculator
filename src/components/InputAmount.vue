<script lang="ts">
import { debounce, roundNum } from '../helpers'
import { defineComponent } from 'vue'
import { i18n } from '../i18n'

export default defineComponent({
	props: {
		value: {
			type: Number,
			required: true,
		},
		disabled: {
			type: Boolean,
		},
	},

	emits: ['value-change'],

	data() {
		return {
			domValue: String(this.value),
			valid: true,
		}
	},

	computed: {
		debouncedInputValueChanged() {
			return debounce(this.inputValueChanged, 500)
		},

		disabledValue() {
			let digits = 10
			if (this.value > 1) {
				digits = 4
			} else if (this.value > 1e-2) {
				digits = 5
			} else if (this.value > 1e-4) {
				digits = 8
			}

			return this.value.toLocaleString(undefined, { maximumFractionDigits: digits })
		},

		usedValue() {
			return this.disabled ? this.disabledValue : this.domValue
		},
	},

	/*
		Why are we using a data property for storing the value and update it
		when the props value changes, instead of just using `this.value`?

		In Vue, components get updated whenever their component data changes.

			In this case, this means that when the value of the input in the
		DOM is not the same as the value in the props (e.g. when the DOM value
		is invalid, and thus gets ignored by the component), the DOM value will
		get overwritten by the 'real' value from the props while updating.

			Unfortunately, this will also delete the most recent user input,
		since that input does not yet exist in the props.

			To avoid this from happening, we use a data property that is
		always updated to be equal to the DOM value, and will only get
		overwritten by the props value when `this.value` changes, and not when
		any other part of the component changes.
	 */
	watch: {
		value(newValue: number) { this.domValue = String(newValue) },
	},

	methods: {
		/** Updates the validity status of the input */
		changeValidity(target: HTMLInputElement, msg: string) {
			target.setCustomValidity(msg)
			target.reportValidity()
			this.valid = msg === ''
		},

		/** Checks the validity of the input field and emits the new value to the parent component */
		inputValueChanged(event: Event) {
			const target = event.target as HTMLInputElement
			this.domValue = target.value

			const valueStr = this.domValue.replace(/,/g, '')

			/* Don't invalidate yet - the user might enter a
				digit, making the input a valid simple fraction */
			if (`${parseFloat(valueStr)}/` === valueStr) {
				console.log('Ignoring incomplete simple fraction')
				return
			}

			const fractionMatchArray = valueStr.match(/([\d.]+) ?\/ ?([\d.]+)/)
			let value = +valueStr

			if (fractionMatchArray) {
				// Handle simple fractions
				value = +fractionMatchArray[1] / +fractionMatchArray[2]
				value = roundNum(value, 5)
			}

			if (!Number.isFinite(value)) {
				this.changeValidity(target, i18n.t('invalidNumber'))
				console.log('Ignoring invalid number input')
				return
			}

			this.$emit('value-change', value)
			this.changeValidity(target, '')
		},
	},
})
</script>

<template>
	<input
		:class="{ 'input-amount': true, 'invalid': !valid }"
		:disabled="disabled"
		:value="usedValue"
		@input="debouncedInputValueChanged"
	>
</template>
