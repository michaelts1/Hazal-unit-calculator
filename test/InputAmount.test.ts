import { beforeEach, describe, expect, it, vi } from 'vitest'
import InputAmount from '../src/components/InputAmount.vue'
import { mount } from '@vue/test-utils'
import { setInvalidValue } from './test-utils'

describe('Component InputAmount, enabled', () => {
	it('exists', () => {
		expect(InputAmount).toBeTruthy()
	})

	let wrapper = mount(InputAmount, { props: { value: 0 } })
	let input = wrapper.find('input:not([disabled])')
	beforeEach(() => { // Reset component
		vi.useFakeTimers()
		wrapper = mount(InputAmount, { props: { value: 0 } })
		input = wrapper.find('input:not([disabled])')

		return vi.useRealTimers // AfterEach function
	})

	it('has an active input field with correct classes', () => {
		expect(input.exists()).toBe(true)
		expect(input.classes()).contain('input-amount')
		expect(input.classes()).not.contain('invalid')
	})

	it('reacts to valid value changes', async () => {
		// Yes, dom value is always a string
		expect(wrapper.vm.domValue).toBe('0')

		await input.setValue(1)
		vi.runOnlyPendingTimers()
		expect(wrapper.vm.domValue).toBe('1')
		expect(input.classes()).not.contain('invalid')

		// Should be registered as input events
		const inputEvents = wrapper.emitted('input')
		expect(inputEvents).toHaveLength(1)

		// Should be registered as value-change event
		const valueChangeEvents = wrapper.emitted('value-change')
		expect(valueChangeEvents).toEqual([ [1] ])
	})

	it('doesn\'t react to invalid value changes', async () => {
		await setInvalidValue(input, 'invalid value')
		vi.runOnlyPendingTimers()
		expect(wrapper.vm.domValue).toBe('invalid value')
		expect(input.classes()).contain('invalid')

		// Should be registered as input events
		const inputEvents = wrapper.emitted('input')
		expect(inputEvents).toHaveLength(2) // Two event were emitted due to `setInvalidValue`

		// Shouldn't be registered as value-change event
		const valueChangeEvents = wrapper.emitted('value-change')
		expect(valueChangeEvents).toBe(undefined)
	})

	it('updates domValue when value prop changes', async () => {
		await wrapper.setProps({ value: 5 })
		expect(wrapper.vm.domValue).toBe('5')
	})

	it('should wait for the user to complete a simple fraction', async () => {
		await setInvalidValue(input, '5/')
		vi.runOnlyPendingTimers()
		expect(wrapper.vm.domValue).toBe('5/')
		expect(input.classes()).not.contain('invalid')

		await input.setValue('5/2')
		vi.runOnlyPendingTimers()
		const valueChangeEvents = wrapper.emitted('value-change')
		expect(valueChangeEvents).toEqual([ [2.5] ])
	})
})

describe('Component InputAmount, disabled', () => {
	const wrapper = mount(InputAmount, {
		props: {
			value: 0,
			disabled: true,
		},
	})

	const input = wrapper.find('input[disabled]')

	it('has a disabled input field with correct classes', () => {
		expect(input.exists()).toBe(true)
		expect(input.classes()).contain('input-amount')
		expect(input.classes()).not.contain('invalid')
		expect(input.attributes()).toHaveProperty('disabled')
	})

	it("doesn't reacts value changes", async () => {
		expect(wrapper.vm.domValue).toBe('0')

		await input.setValue(5)
		expect(wrapper.vm.domValue).toBe('0')
	})

	it('Rounds displayed value according the size of the value', async () => {
		await wrapper.setProps({ value: 1.555_555_555_555 })
		expect(wrapper.vm.disabledValue).toBe('1.5556')

		await wrapper.setProps({ value: 0.555_555_555_555 })
		expect(wrapper.vm.disabledValue).toBe('0.55556')

		await wrapper.setProps({ value: 0.005_555_555_555 })
		expect(wrapper.vm.disabledValue).toBe('0.00555556')

		await wrapper.setProps({ value: 0.000_055_555_555 })
		expect(wrapper.vm.disabledValue).toBe('0.0000555556')
	})
})
