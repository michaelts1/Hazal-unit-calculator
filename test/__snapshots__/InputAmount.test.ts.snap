// Vitest Snapshot v1

exports[`Component InputAmount, disabled > matches snapshot 1`] = `"<input class=\\"conversion-input\\" disabled=\\"\\" data-v-519321c2=\\"\\">"`;

exports[`Component InputAmount, enabled > matches snapshot 1`] = `"<input class=\\"conversion-input\\" data-v-519321c2=\\"\\">"`;
